corehq/motech/models.py

```diff
@@ -26,56 +43,155 @@ class ConnectionSettings(models.Model):
+        if self.last_token_aes:
+            plaintext = b64_aes_decrypt(self.last_token_aes)
+            return json.loads(plaintext)
+        return None
+
+    @last_token.setter
+    def last_token(self, token: Optional[dict]):
+        if token is None:
+            self.last_token_aes = ''
+        else:
+            plaintext = json.dumps(token)
+            self.last_token_aes = b64_aes_encrypt(plaintext)
 
     @property
     def notify_addresses(self):
         return [addr for addr in re.split('[, ]+', self.notify_addresses_str) if addr]
 
-    def get_requests(self, payload_id, logger):
+    def get_requests(
+        self,
+        payload_id: Optional[str] = None,
+        logger: Optional[Callable] = None,
+    ):
         from corehq.motech.requests import Requests
+
+        auth_manager = self.get_auth_manager()
         return Requests(
             self.domain,
             self.url,
-            self.username,
-            self.plaintext_password,
             verify=not self.skip_cert_verify,
+            auth_manager=auth_manager,
             notify_addresses=self.notify_addresses,
             payload_id=payload_id,
             logger=logger,
-            auth_type=self.auth_type,
         )
 
+    def get_auth_manager(self):
+        if self.auth_type is None:
+            return AuthManager()
+        if self.auth_type == BASIC_AUTH:
+            return BasicAuthManager(
+                self.username,
+                self.password,
+            )
+        if self.auth_type == DIGEST_AUTH:
+            return DigestAuthManager(
+                self.username,
+                self.password,
+            )
+        if self.auth_type == OAUTH1:
+            return OAuth1Manager(
+                client_id=self.client_id,
+                client_secret=self.client_secret,
+                api_endpoints=self.get_oauth1_api_endpoints(),
+                last_token=self.last_token,
+            )
+        if self.auth_type == BEARER_AUTH:
+            return BearerAuthManager(
+                self.username,
+                self.password,
+            )
+        if self.auth_type == OAUTH2_PWD:
+            return OAuth2PasswordGrantManager(
+                self.url,
+                self.username,
+                self.password,
+                client_id=self.client_id,
+                client_secret=self.client_secret,
+                api_settings=self.get_oauth2_api_settings(),
+                last_token=self.last_token,
+            )
+
+    def get_oauth1_api_endpoints(self):
+        if self.api_auth_settings in dict(oauth1_api_endpoints):
+            return getattr(corehq.motech.auth, self.api_auth_settings)
+        raise ValueError(_(
+            f'Unable to resolve API endpoints {self.api_auth_settings!r}. '
+            'Please select the applicable API auth settings for the '
+            f'{self.name!r} connection.'
+        ))
+
+    def get_oauth2_api_settings(self):
```

Comment:
looks like this and `get_oauth1_api_endpoints` are only used in this class, should they be marked private?

Reply:
9b889fb